File: pe005/pe005.py
#!/usr/bin/python

"""
Project Euler, Problem 5:

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

https://projecteuler.net/problem=5
"""

from functools import reduce

max_number = 20

def get_gcd2(a, b):
    while b > 0:
        b, a = a % b, b
    return a

def get_gcd(numbers):
    return reduce(get_gcd2, numbers)

def get_lcm2(a, b):
    return (a * b) / get_gcd2(a, b)

def get_lcm(numbers):
    return reduce(get_lcm2, numbers)

def main():
    print(int(get_lcm(range(1, max_number + 1))))

if __name__ == "__main__":
    main()

File: pe005/pe005.go
/*
Project Euler, Problem 5:

What is the smallest positive number that is evenly divisible by all of
the numbers from 1 to 20?

https://projecteuler.net/problem=5
*/

package main

import "fmt"

const max = 20

type reduceFunc func(int, int) int

func reduce(f reduceFunc, numbers []int) int {
	result := numbers[0]
	for _, n := range numbers {
		result = f(result, n)
	}
	return result
}

func getGCD2(a, b int) int {
	for a > 0 {
		a, b = b%a, a
	}
	return b
}

func getLCM2(a, b int) int {
	return int(a * b / getGCD2(a, b))
}

func getGCD(numbers []int) int {
	return reduce(getGCD2, numbers)
}

func getLCM(numbers []int) int {
	return reduce(getLCM2, numbers)
}

func main() {
	var numbers []int
	for i := 2; i <= max; i++ {
		numbers = append(numbers, i)
	}
	fmt.Println(getLCM(numbers))
}
